provider/oracle/provider.go

```diff
@@ -0,0 +1,175 @@
+// passed through PrepareConfig at some point in its lifecycle.
+//
+// This operation is not performing any expensive operation.
+func (e *environProvider) Open(params environs.OpenParams) (environs.Environ, error) {
+	logger.Debugf("opening model %q", params.Config.Name())
+	if err := e.checkSpec(params.Cloud); err != nil {
+		return nil, errors.Annotatef(err, "validating cloud spec")
+	}
+
+	logger.Debugf("creating the oracle client for %q cloud", params.Cloud.Name)
+	cli, err := oci.NewClient(oci.Config{
+		Username: params.Cloud.Credential.Attributes()["username"],
+		Password: params.Cloud.Credential.Attributes()["password"],
+		Endpoint: params.Cloud.Endpoint,
+		Identify: params.Cloud.Credential.Attributes()["identity-domain"],
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	e.client = cli
+
+	environ := newOracleEnviron(e, params)
+	return environ, nil
+}
+
+// Validate method will validate model configuration
+// This will ensure that the config passed is a valid configuration for the oracle cloud.
+// If old is not nil, Validate should use it to determine whether a configuration change is valid.
+func (e environProvider) Validate(cfg, old *config.Config) (valid *config.Config, _ error) {
+	return nil, nil
+}
+
+// CredentialSchemas returns credential schemas, keyed on authentication type. This is used to validate existing oracle credentials, or to generate new ones.
+func (e environProvider) CredentialSchemas() map[cloud.AuthType]cloud.CredentialSchema {
+	return map[cloud.AuthType]cloud.CredentialSchema{
+		cloud.UserPassAuthType: {{
+			"username", cloud.CredentialAttr{
+				Description: "account username",
+			},
+		}, {
+			"password", cloud.CredentialAttr{
+				Description: "account password",
+				Hidden:      true,
+			},
+		}, {
+			"identity-domain", cloud.CredentialAttr{
+				Description: "indetity domain of the oracle account",
+			},
+		}},
+	}
+}
+
+// DetectCredentials automatically detects one or more oracle credentials from the environmnet. This may involve, for example inspecting environmnet variables, or reading configuration files in well-defined locations.
+// If no credentials can be detected, the func will return an error satisfying errors.IsNotFound
+func (e environProvider) DetectCredentials() (*cloud.CloudCredential, error) {
+	return nil, errors.NotFoundf("credentials")
+}
+
+// FinalizeCredential finalizes a oracle credential, updating any attributes as
+// necessarry. This is done clinet-side, when adding the credential to credentials.yaml
+// and before uploading crdentials to the controller.
+// The provider may completely alter a credential, even going as far as changing the auth-type,
+// but the output must be a fully formed credential that is orcale complaint.
+func (e environProvider) FinalizeCredential(
+	cfx environs.FinalizeCredentialContext,
+	params environs.FinalizeCredentialParams,
+) (*cloud.Credential, error) {
+	_, ok := params.Credential.Attributes()["identity-domain"]
```

Comment:
This method is not the place to validate credentials, since it's only called when adding new credentials. Please see "validateCloudSpec" in other providers, which should be called by EnvironProvider.Open and EnvironProvider.PrepareConfig methods.